crates/sui-types/src/crypto.rs

```diff
@@ -1,205 +1,109 @@
 // Copyright (c) 2022, Mysten Labs, Inc.
 // SPDX-License-Identifier: Apache-2.0
-use crate::base_types::{decode_bytes_hex, AuthorityName, SuiAddress};
+use crate::base_types::{AuthorityName, SuiAddress, ToAddress};
 use crate::committee::{Committee, EpochId};
 use crate::error::{SuiError, SuiResult};
 use crate::sui_serde::Base64;
 use crate::sui_serde::Readable;
 use crate::sui_serde::SuiBitmap;
-use anyhow::anyhow;
 use anyhow::Error;
 use base64ct::Encoding;
 use digest::Digest;
-use ed25519_dalek as dalek;
-use ed25519_dalek::{Keypair as DalekKeypair, Verifier};
-use narwhal_crypto::ed25519::{Ed25519KeyPair, Ed25519PublicKey};
-use once_cell::sync::OnceCell;
+use narwhal_crypto::ed25519::{
+    Ed25519KeyPair, Ed25519PrivateKey, Ed25519PublicKey, Ed25519PublicKeyBytes,
+};
+pub use narwhal_crypto::traits::KeyPair as KeypairTraits;
+pub use narwhal_crypto::traits::{
+    AggregateAuthenticator, Authenticator, SigningKey, ToFromBytes, VerifyingKey,
+};
+use narwhal_crypto::Verifier;
 use rand::rngs::OsRng;
 use roaring::RoaringBitmap;
 use schemars::JsonSchema;
 use serde::{Deserialize, Serialize};
 use serde_with::serde_as;
 use serde_with::Bytes;
 use sha3::Sha3_256;
-use std::borrow::Borrow;
 use std::collections::HashMap;
 use std::hash::{Hash, Hasher};
-use std::str::FromStr;
 
-// TODO: Make sure secrets are not copyable and movable to control where they are in memory
-#[derive(Debug)]
-pub struct KeyPair {
-    key_pair: DalekKeypair,
-    public_key_cell: OnceCell<PublicKeyBytes>,
```

Comment:
The main reason for this existence of this memoizer pattern is to to avoid converting to bytes each time.

Reply:
I will send in a PR to Narwhal that has these changes!

Reply:
There is no need for this memoizer pattern on Ed25519. 

Here's a dalek `PublicKey`: https://github.com/dalek-cryptography/ed25519-dalek/blob/main/src/public.rs#L40
the `CompressedEdwards`: https://github.com/dalek-cryptography/curve25519-dalek/blob/main/src/edwards.rs#L164

The fact that a `Foo` implements `AsRef<[u8]>` exactly represents that you can get a zero-copy reference to its byte representation.

`dalek_key_pair.public.as_ref::<[u8]>()` involves exactly zero byte conversions.

Reply:
This abstraction is done in the narwhal library :)